README
This is the placeholder of the README file.
CPANTS http://cpants.perl.org/ wants it to be here.

Anyway, the module is a Gtk2 based wrapper of the ack command 
line grep replacement.


CHANGES
Changes of App::Wack

0.05 2008.03.27

 Add history of searches (within the session) and show them in a drop-down list.
 Show pop-up error messages if invalid regex given
 When selecting specific file make sure it is acked even if that file would 
   not be searched because of its type. So searching on single file explicitly
   sets the -a flag.
 Start using App::Ack directly and not via its command line interface
 Save/load configuration (and history) (File::HomeDir, YAML)
 Move the signal handles to App::Wack::Signals and keep the application in App::Wack

0.04 2008.02.27

 Enable selecting files in the directory and file chooser
 Add several more options including -A and -B context

0.03 2007.08.19

 Add directory chooser to search strings in a selected directory

0.02 2007.08.18

Add several options available on the command line as checkboxes on the app.
 Ignore case
 Invert Match
 Word regexp
 Quote literal
 No subdirs

0.01 2007.08.18

First version released


Makefile.PL
# Note: this file was auto-generated by Module::Build::Compat version 0.03
use ExtUtils::MakeMaker;
WriteMakefile
(
          'NAME' => 'App::Wack',
          'VERSION_FROM' => 'lib/App/Wack.pm',
          'PREREQ_PM' => {
                           'App::Ack' => '1.78',
                           'Carp' => '0',
                           'Cwd' => '0',
                           'File::HomeDir' => '0',
                           'File::Spec' => '0',
                           'Gtk2' => '1.080',
                           'Gtk2::GladeXML' => '0',
                           'List::Util' => '0',
                           'Test::More' => '0',
                           'Test::Pod' => '0',
                           'Test::Pod::Coverage' => '0',
                           'YAML' => '0'
                         },
          'INSTALLDIRS' => 'site',
          'EXE_FILES' => [
                           'bin/wack'
                         ],
          'PL_FILES' => {}
        )
;


lib/App/Wack.pm
package App::Wack;
use strict;
use warnings;

use Gtk2 -init;
use Gtk2::Helper;

use App::Ack;

use File::HomeDir ();
use YAML          ();
use File::Spec    ();
use Data::Dumper  qw(Dumper);
use List::Util    qw(max min);

use App::Wack::Glade;

sub say { print @_, "\n";}

our $VERSION = '0.05';

my @content;

my $tag;
my $config;

sub setup {
    read_config();
    set_defaults();
    return;
}

sub _get_config_file {
    my $home  = File::HomeDir->my_home;
    return File::Spec->catfile($home, 'wack.yml');
}
sub save_config {
    my $file = _get_config_file();
    YAML::DumpFile($file, $config);
    return;
}

sub read_config {
    my $file = _get_config_file();
    if (-e $file) {
        $config = YAML::LoadFile($file);
    }

    return;
}
sub set_defaults {
    $config->{conf}{max_search_history} ||= 10;
    $config->{recent_searches}          ||= [];
}


sub get_config {
    return $config;
}

sub get_help {
    my $f = join "\n",  App::Ack::filetypes_supported();
    # TODO: get the copyright from ack automatically
    #my $ack_copyrioght = App::Ack::version_statement($COPYRIGHT);
    #my $ack_copyright = App::Ack::get_copyright();
    my $thppt = App::Ack::_get_thpppt();

    my $ack_version = $App::Ack::VERSION;
    my $warning = '';
    my $dev_version = '1.78';
    if ($ack_version ne $dev_version) {
        $warning =  "\nThis version of Wack was developed with ack version $dev_version.\n";
        $warning .= "As you are currently using version $ack_version, there might be\n";
        $warning .= "some issues related to the version mismatch\n";
    }
    my $help = <<"END_HELP";
This is version $VERSION of wack
Using ack version $App::Ack::VERSION
Gtk version: $Gtk2::VERSION
Perl version: $]

wack, Copyright 2007-2008 Gabor Szabo http://search.cpan.org/dist/App-Wack
ack,  Copyright 2005-2008 Andy Lester http://search.cpan.org/dist/ack
Gtk2, Copyright 2003-2008 the gtk2-perl team http://search.cpan.org/dist/Gtk2
perl, Copyright 1987-2008 Larry Wall http://search.cpan.org/dist/perl

$thppt

$warning

There are currently two ways of using Wack.

Ack mode:
Select a directory (or a file) by pressing the Browse button.
Type in a search string or regular expression.
Select one or more of the checkboxes on the left and select the
before and after context values.
Click on search.


In-memory mode:
Select a single file by pressing the Browse button.
Check the "In memory" checkbox. The files should be loaded in the viewer.
Type in a search string or regular expression and check (or uncheck) the options.
Currently supporting the
Invert Match, Ignore Case checkboxes and the before and after context selectors.
The matching lines should be displayed as you make your changes.


END_HELP

    return $help;
}

sub _update_regex_history {
    my ($regex) = @_;

    my @recent_searches = @{ $config->{recent_searches} };
    if ($regex) {
        my @temp = grep {$regex ne $_} @recent_searches;
        if (@temp != @recent_searches) {
            @recent_searches = @temp;
        } elsif (@recent_searches >= $config->{conf}{max_search_history}) {
            shift @recent_searches;
        }
        push @recent_searches, $regex;
    }
    $config->{recent_searches} = \@recent_searches;
    return \@recent_searches;
}

sub _validate_regex {
    my ($regex) = @_;
    if (not $config->{opts}{Q}) {    # literal
        eval "qr/$regex/";           ## no critic
        if ($@) {
            (my $error = $@) =~ s/ at \(eval \d+\).*//;
            die "$error\n";
        }
    }
    return 1;
}

   
#sub directory_changed_cb {
#    exit;
#}

# see t/module.t in ack distro
sub fill_type_wanted {
    for my $i ( App::Ack::filetypes_supported() ) {
        $App::Ack::type_wanted{ $i } = undef;
    }
}


sub ack {
    my ($regex, $dir) = @_;
    #say $regex;
    #say $dir;
#    $gladexml->get_widget('stop-button')->activate;

#    my $count = 0;
    #$gladexml->get_widget('stop-button')->set('sensitive', 1);
#    $gladexml->get_widget('stop-button')->sensitive(1);
    #return;

#   if (open my $ACK, "ack $str $dir |") { ## no critic
#        $tag = Gtk2::Helper->add_watch( fileno( $ACK ), in => sub {
#            if ( $stop_now or eof( $ACK ) ) {
#                close( $ACK );
#                #$gladexml->get_widget('stop-button')->set('sensitive', 0);
#                Gtk2::Helper->remove_watch( $tag );
#            } else {
#                my $line = <$ACK>;
#                $cb->($line);
#            }
#            return 1;
#        } );
#
#    } else {
#        show_error("Could not find the ack executable.");
#    }
#    return;

    my $config = get_config();
    my %opts = %{ $config->{opts} };

    $opts{regex} = $regex;
    if (-f $dir) {
        $opts{all} = 1;
    }
#    print Dumper \%opts;

    my $what = App::Ack::get_starting_points( [$dir], \%opts );
    fill_type_wanted();
#    $App::Ack::type_wanted{cc} = 1;

    #$opts{show_filename} = 1;
#    $opts{follow} = 0;
    my $iter = App::Ack::get_iterator( $what, \%opts );
    App::Ack::filetype_setup();
    App::Ack::print_matches( $iter, \%opts );
}


sub search {
    my ($search) = @_;
    my $opts = get_options();
    my $modifiers = '' . ($opts->{i} ? 'i' : '');
    my $invert = $opts->{v};
    
    my $regex = eval "qr/$search/$modifiers"; ## no critic
    # turn on some visible but not inturisve flag that indicates the current regex is invalid?
    die $@ if $@;

    my @selected;
    my $last_line;
    for my $i (0..@content-1) {
        my $add = $invert ? $content[$i] !~ /$regex/ : $content[$i] =~ /$regex/;
        if ($add) {
            for my $j ( max(0, $i - $opts->{before_context}) .. min($#content, $i + $opts->{after_context}) ) {
                if (not defined $last_line or $j > $last_line) {
                    push @selected, "$j: $content[$j]";
                    $last_line = $j;
                }
            }
        }
    }
    return join '', @selected;
}

sub get_options {
    my $config = get_config();
    return $config->{opts};
}

sub set_option {
    my ($name, $value) = @_;

    if (defined $value and $value ne '') {
        $config->{opts}{$name} = $value;
    } else {
        delete $config->{opts}{$name};
    }

    return;
}


sub read_file {
    my ($file) = @_;
    open my $fh, '<', $file or return;

    @content = <$fh>;
    return join '', @content;
}



=head1 NAME

App::Wack - the wisual ack

=head1 SYNOPSIS

You should only use the 

 wack

command and not the module itself.

=head1 DESCRIPTION

App::Wack is the implementation of the Gtk2 and GladeXML based 
GUI around L<ack>, the grep-like text finder.

=head1 DEVELOPMENT

Subversion repository:
L<http://svn1.hostlocal.com/szabgab/trunk/wack/>

Bugs L<https://rt.cpan.org/> 
Email: bug-app-wack to rt.cpan.org

=head1 TODO

add more of the available flags as checkboxes

add the selected file type list as a configuration option
File selection available in ack:

 -  default
 -a all files
    
Show some sign that search is still in progress (and/or that search ended) 

Allow the user to abort search in the middle

Highlight the matches.

Allow loading a file into memory and allow "work as I type the regex" mode meaning
the results are updated after every change in the regex.

Remember previous regexes
Remember previously search file and directory names

Bug: If the file/directory window is closed by ESC or by clicking on the x 
it cannot be opened again.

=head1 COPYRIGHT

Copyright 2007-2008 Gabor Szabo, all rights reserved.

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

=cut

1;



lib/App/Wack/Signals.pm
package App::Wack::Signals;
use strict;
use warnings;

use Carp          qw(croak);
use Cwd           qw(cwd);
use Data::Dumper  qw(Dumper);

use Gtk2::GladeXML;

our $VERSION = '0.01';

sub say { print @_, "\n";}

my $gladexml;

{
    no warnings 'redefine';
    sub App::Ack::print_first_filename { _add_results("$_[0]\n"); }
    sub App::Ack::print_separator      { _add_results("--\n"); }
    sub App::Ack::print                { _add_results($_[0]); }
    sub App::Ack::print_filename       { _add_results("$_[0]$_[1]"); }
    sub App::Ack::print_line_no        { _add_results("$_[0]$_[1]"); }
}

my %options = (
    'check-case-sensitive' => 'i',
    'check-invert-match'   => 'v',
    'check-word-regexp'    => 'w',
    'check-literal'        => 'Q',
    'check-subdirs'        => 'n',
    'check-all-types'      => 'all',
    'before_context'       => 'before_context',
    'after_context'        => 'after_context',
);

AUTOLOAD {
    our $AUTOLOAD;
    if ($ENV{WACK_TESTING}) {
        croak "Missing function '$AUTOLOAD' called";
    } else {
        print $AUTOLOAD . " <-\n";
    }
}
sub DESTROY {
}


sub setup {
    $gladexml = Gtk2::GladeXML->new_from_buffer(App::Wack::Glade::get_xml());
    $gladexml->signal_autoconnect_from_package("App::Wack::Signals");

    return;
}

sub run {
    my ($config) = @_;
    my $recent_searches = $config->{recent_searches};
    _set_options($config->{opts} );
    _update_regex_history_box( $recent_searches );
    $gladexml->get_widget('directory')->set_text(cwd);
    Gtk2->main;
}


sub on_main_destroy {
    _close_app();
}
sub on_quit_clicked {
    _close_app();
}

sub _close_app {
	Gtk2->main_quit;
    App::Wack::save_config();
    exit;
}

sub _set_options {
    my ($opt) = @_;

    foreach my $widget (keys %options) {
        if ($widget eq 'before_context' or $widget eq 'after_context') {
            my $combobox = $gladexml->get_widget($widget);
            foreach (0..4) {
                $combobox->append_text ($_);
            }
            $combobox->set_active($opt->{ $options{$widget} } || 0);
        } else {
            if ($opt->{ $options{$widget} } ) {
                $gladexml->get_widget($widget)->set_active(1);
            }
        }
    }
}


sub on_check_toggled {
    my ($button) = @_;
    my $key = $options{ $button->get_name };
    App::Wack::set_option($key, $button->get_active);

    $gladexml->get_widget('search-text')->grab_focus;

    run_search();

    return;
}
sub on_combo_changed {
    return on_check_toggled(@_);
}


sub _clean_results {
    my $result = $gladexml->get_widget('results')->get_buffer;
    $result->delete($result->get_start_iter, $result->get_end_iter);
    return;
}
sub _set_result {
    my ($text) = @_;

    _clean_results();
    my $result = $gladexml->get_widget('results')->get_buffer;
    $result->set_text($text);  
    return;
}

sub on_check_inmemory_toggled {
    return if not $gladexml->get_widget('check-inmemory')->get_active;
    my $file = $gladexml->get_widget('directory')->get_text();
    # TODO disable the irrelevant buttons and checkboxes?
    # or shall I create a separate window with the available options for the
    # in-memory tool?
    if (not -f $file) {
        show_error("First select a file please");
        $gladexml->get_widget('check-inmemory')->set_active(0);
        return;
    }
    _clean_results();

    if (my $content = App::Wack::read_file($file)) {
        _set_result($content);
    } else {
        show_error("Could not open file $!");
    };

    # TODO turn off the click box if a non-file is selected? What if we are editing the path to get to a new file?
 
    return;
}

sub on_changed_search {
    run_search();
    return;
}

sub run_search {
    return if not $gladexml->get_widget('check-inmemory')->get_active;
    my $search = $gladexml->get_widget('search-text')->child->get_text;

    eval {
        my $result = App::Wack::search($search);
        App::Wack::Signals::_set_result($result);
    };
    if ($@) {
        show_error("regex problem: $@");
    }

    # TODO: highlight
    return;
}

sub _update_regex_history_box {
    my ($searches) = @_;
    my $search_widget = $gladexml->get_widget('search-text');
    $search_widget->get_model->clear;
    foreach my $txt (@$searches) {
        $search_widget->prepend_text($txt);
    }
    return;
}


sub on_search_clicked {
    #my $result   = $gladexml->get_widget('results')->get_buffer;
    my $regex    = $gladexml->get_widget('search-text')->child->get_text;
    my $dir      = $gladexml->get_widget('directory')->get_text();
    #say $regex;
    eval {
        App::Wack::_validate_regex($regex);
    };
    if ($@) {
        show_error("Invalid regex: $@");
        return;
    }

    my $searches = App::Wack::_update_regex_history($regex);
    _update_regex_history_box($searches);
    _clean_results();

    App::Wack::ack($regex, $dir);

    #$opt{m} = $MAX_ROWS;

    return;
}
 
sub on_config_clicked {
    warn "Not implemented yet\n";
}


sub show_error {
    my ($msg) = @_;
#    print "ERROR $msg\n";
    my $dialog = Gtk2::MessageDialog->new ($gladexml->get_widget('main'),
                                             'destroy-with-parent',
                                             'error',
                                             'ok',
                                             $msg);
    my $response = $dialog->run;
    $dialog->destroy;
    return;
}

sub on_browse_button_clicked {
    my $dir    = $gladexml->get_widget('directory')->get_text();
    if (-d $dir) {
        $gladexml->get_widget('directory-selector')->set_current_folder($dir);
    }
    $gladexml->get_widget('directory-selector')->show;
    return; 
}


sub on_directory_window_closed {
    #print "1\n";
    #if ($gladexml->get_widget('directory-selector')) {
        #$gladexml->get_widget('directory-selector')->hide;
    #}
    return;
}
sub on_directory_selector_deleted {
    # when directory closed with escape or x
    # but it still gets destroyed
    return;
}


sub on_directory_ok_clicked {
    my $filename = $gladexml->get_widget('directory-selector')->get_filename;
    $gladexml->get_widget('directory')->set_text($filename);
    $gladexml->get_widget('directory-selector')->hide;
    return;
}

sub on_directory_cancel_clicked {
    $gladexml->get_widget('directory-selector')->hide;
    return;
}

sub _add_results {
    my ($text) = @_;
#say $text;
    my $result = $gladexml->get_widget('results')->get_buffer;
    my $end = $result->get_end_iter;
    $result->insert($end, $text);
    #print $text;
    return;
}

sub on_help_clicked {

    my $help = App::Wack::get_help();
    _set_result($help);
    return;
}
sub on_activate {
    on_search_clicked();
}

sub on_search_activate {
    on_search_clicked();
}

sub on_stop_button_clicked {
    print "STOP\n";
    #$stop_now++;
    return;
}

sub get_gladexml {
    return $gladexml;
}

1;




Build.PL
use strict;
use warnings;

use 5.006;

use Module::Build;

# in the development environment only
if (-e 'create_glade') {
    system "$^X -I lib create_glade";
}

my $build = Module::Build->new(
    module_name        => 'App::Wack',
    dist_version_from  => 'lib/App/Wack.pm',
    license            => 'perl',
    create_makefile_pl => 'traditional',
    script_files       => 'bin/wack',
    requires           => {
            'Gtk2'                   => '1.080', 
            'Gtk2::GladeXML'         => '0',
            'App::Ack'               => '1.78',
            'File::HomeDir'          => '0',
            'YAML'                   => '0',

            'Carp'                   => '0',
            'Cwd'                    => '0',
            'File::Spec'             => '0',
            'List::Util'             => '0',
    },
    build_requires     => {
            'Test::More'             => '0',
            'Test::Pod'              => '0',
            'Test::Pod::Coverage'    => '0',
    },
);

$build->create_build_script;

